tests/components/xiaomi_miio/test_config_flow.py

```diff
@@ -119,7 +125,83 @@ async def test_config_flow_gateway_success(hass):
         config_flow.CONF_FLOW_TYPE: config_flow.CONF_GATEWAY,
         CONF_HOST: TEST_HOST,
         CONF_TOKEN: TEST_TOKEN,
-        "gateway_id": TEST_GATEWAY_ID,
         "model": TEST_MODEL,
         "mac": TEST_MAC,
     }
+
+
+async def test_zeroconf_gateway_success(hass):
+    """Test a successful zeroconf discovery of a gateway."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN,
+        context={"source": config_entries.SOURCE_ZEROCONF},
+        data={
+            zeroconf.ATTR_HOST: TEST_HOST,
+            ZEROCONF_NAME: TEST_ZEROCONF_NAME,
+            ZEROCONF_PROP: {ZEROCONF_MAC: TEST_MAC},
+        },
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "gateway"
+    assert result["errors"] == {}
+
+    mock_info = get_mock_info()
+
+    with patch(
+        "homeassistant.components.xiaomi_miio.gateway.gateway.Gateway.info",
+        return_value=mock_info,
+    ), patch(
+        "homeassistant.components.xiaomi_miio.async_setup_entry", return_value=True
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {CONF_NAME: TEST_NAME, CONF_TOKEN: TEST_TOKEN},
+        )
+
+    assert result["type"] == "create_entry"
+    assert result["title"] == TEST_NAME
+    assert result["data"] == {
+        config_flow.CONF_FLOW_TYPE: config_flow.CONF_GATEWAY,
+        CONF_HOST: TEST_HOST,
+        CONF_TOKEN: TEST_TOKEN,
+        "model": TEST_MODEL,
+        "mac": TEST_MAC,
+    }
+
+
+async def test_zeroconf_unknown_device(hass):
+    """Test a failed zeroconf discovery because of a unknown device."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN,
+        context={"source": config_entries.SOURCE_ZEROCONF},
+        data={
+            zeroconf.ATTR_HOST: TEST_HOST,
+            ZEROCONF_NAME: "not-a-xiaomi-miio-device",
+            ZEROCONF_PROP: {ZEROCONF_MAC: TEST_MAC},
+        },
+    )
+
+    assert result["type"] == "abort"
+    assert result["reason"] == "not_xiaomi_miio"
+
+
+async def test_zeroconf_no_data(hass):
+    """Test a failed zeroconf discovery because of no data."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN, context={"source": config_entries.SOURCE_ZEROCONF}, data={}
+    )
+
+    assert result["type"] == "abort"
+    assert result["reason"] == "not_xiaomi_miio"
+
+
+async def test_zeroconf_missing_data(hass):
+    """Test a failed zeroconf discovery because of missing data."""
+    result = await hass.config_entries.flow.async_init(
+        const.DOMAIN,
+        context={"source": config_entries.SOURCE_ZEROCONF},
+        data={zeroconf.ATTR_HOST: TEST_HOST, ZEROCONF_NAME: TEST_ZEROCONF_NAME},
+    )
+
+    assert result["type"] == "abort"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)